ulib/FStar.Calc.fst

```diff
@@ -0,0 +1,41 @@
+module FStar.Calc
+
+open FStar.Preorder (* for `relation` *)
+
+noeq
+type calc_proof #t : list (relation t) -> t -> t -> Type =
+  | CalcRefl : #x:t -> calc_proof [] x x
+  | CalcStep : rs:(list (relation t)) -> #p:(relation t) ->
+               #x:t -> #y:t -> #z:t -> calc_proof rs x y -> squash (p y z) -> calc_proof (p::rs) x z
+
+let rec calc_chain_related (#t : Type) (rs : list (relation t)) (x y : t) : Type0 =
+  match rs with
+  | [] -> x == y
+  | r1::rs -> exists w. calc_chain_related rs x w /\ r1 w y
+
+(* Every chain of `t`'s related by `rs` **(reversed!)** has its endpoints related by p *)
+let calc_chain_compatible (#t : Type) (rs : list (relation t)) (p : relation t) : Type0 =
+  forall x y. calc_chain_related rs x y ==> p x y
+
+let rec elim_calc_proof #t rs (#x #y : t) (pf : calc_proof rs x y)
+    : Lemma (ensures (calc_chain_related rs x y))
+            (decreases pf) =
+  match pf with
+  | CalcRefl -> ()
+  | CalcStep rs #p' #x #y #z pf p'xy -> elim_calc_proof rs pf
+
+let calc_init (#t:Type) (x:t) : calc_proof [] x x = CalcRefl
+
+let calc_step (#t:Type) (#rs : list (relation t)) (#x #y : t)
+         (p : relation t)                 (* Preorder for this step *)
+         (z : t)                          (* Next expression *)
+         (pf : unit -> calc_proof rs x y) (* Rest of the proof *)
+         (j : unit -> squash (p y z))     (* Justification, thunked to avoid confusion such as #1397 *)
+         : Tot (calc_proof (p::rs) x z)
+         (* Need to annotate #p seemingly due to #1486 *)
+         = CalcStep rs #p (pf ()) (j ())
+
+let calc_finish (#t:Type) (#rs : list (relation t)) (p : relation t) (#x #y :t) (pf : calc_proof rs x y)
+  : Lemma (requires (calc_chain_compatible rs p))
```

Comment:
Curious to understand how this works. To satisfy the precondition of the lemma, how is `calc_chain_compatible` proven? As in, would it involve unfolding of the recursive definition `calc_chain_related` by Z3 (and thus constrained by the fuel setting)?

Reply:
Ah, yes, I forgot I wanted to add a `normalize_term` to this so the unfolding would happen in the normalizer, leaving Z3 only with a quantification to be proven. Without that, large chains do fail. I'm playing with that now.

Reply:
This pattern may be useful: https://github.com/project-everest/hacl-star/commit/7623202eb380db4abaf13121214446b9bd04ce8d#r31557851

Reply:
You can also mark `calc_chain_compatible` as `opaque_to_smt`.